File: PrimaMateria/PrimaMateria/Classes/ElementButton.m
//
//  ElementButton.m
//  PrimaMateria
//
//  Created by Jerry Porter on 1/5/2011.
//  Copyright 2016 xTrensa. All rights reserved.
//

@interface ElementButton()

-(void)colorChanged: (NSNotification *)aNotification;

@end

@implementation ElementButton

-(void)colorChanged: (NSNotification *)aNotification {
    [self setupColors];
    [self setNeedsDisplay];
}

- (void)setupColors {
    UIColor *backgroundColor = [[XTRDataSource sharedInstance] elementAtIndex: self.tag].seriesColor;
    UIColor *selectedColor = [UIColor reverseColorOf: backgroundColor];
    UIImage *backgroundImage = [UIImage imageWithColor: backgroundColor size: self.frame.size];
    UIImage *selectedImage = [UIImage imageWithColor: selectedColor size: self.frame.size];
    
    [self setBackgroundImage: backgroundImage forState: UIControlStateNormal];
    [self setBackgroundImage: selectedImage forState: UIControlStateHighlighted];
    [self setBackgroundImage: selectedImage forState: UIControlStateSelected];
}

- (void)awakeFromNib {
	[super awakeFromNib];
    
    self.backgroundColor = UIColor.clearColor;
    
    [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(colorChanged:) name: NOTIFICATION_SERIES_COLOR_CHANGED object: nil];
	[self setTitleColor: [[[XTRDataSource sharedInstance] elementAtIndex: self.tag] standardConditionColor] forState: UIControlStateNormal];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver: self name: NOTIFICATION_SERIES_COLOR_CHANGED object: nil];
}

@end

File: PrimaMateria/PrimaMateria/ElementButton.swift
//
//  ElementButton.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/14/16.
//  Copyright © 2016 xTrensa. All rights reserved.
//

@objc class ElementButton : GeneralButton {
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder:aDecoder)
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        let element : XTRElement = XTRDataSource.sharedInstance().elementAtIndex(UInt(self.tag))

        self.backgroundColor = UIColor.clearColor()
        self.setTitleColor(element.standardConditionColor(), forState: UIControlState.Normal)
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(ElementButton.colorChanged(_:)), name: NOTIFICATION_SERIES_COLOR_CHANGED, object: nil)
    }
    
    // MARK: - Notification Methods
    
    func colorChanged(aNotification: NSNotification) {
        self.setupColors()
        self.setNeedsDisplay()
    }
    
    // MARK: - Internal Methods

    override func setupColors() {
        let element : XTRElement = XTRDataSource.sharedInstance().elementAtIndex(UInt(self.tag))
        let backgroundColor : UIColor = element.seriesColor()
        let selectedColor : UIColor = backgroundColor.inverseColor()
        let backgroundImage : UIImage = UIImage.imageFromColor(backgroundColor, andSize: self.frame.size)
        let selectedImage : UIImage = UIImage.imageFromColor(selectedColor, andSize: self.frame.size)
        self.setBackgroundImage(backgroundImage, forState: UIControlState.Normal)
        self.setBackgroundImage(selectedImage, forState: UIControlState.Highlighted)
        self.setBackgroundImage(selectedImage, forState: UIControlState.Selected)
    }
    
    // MARK: - Memory Management Methods

    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self, name: NOTIFICATION_SERIES_COLOR_CHANGED, object: nil)
    }
}
